src/driver/compile.ml

```diff
@@ -111,76 +114,62 @@ let compile ?partial ~partial_dir (all : Odoc_unit.any list) =
-      let units = try Util.StringMap.find hash hashes with _ -> [] in
-      let r =
+    let compile_one compile_other (unit : Odoc_unit.intf Odoc_unit.t) =
+      let (`Intf { Odoc_unit.deps; _ }) = unit.kind in
+      let _fibers =
         Fiber.List.map
-          (fun unit ->
-            match
-              Hashtbl.find_opt tbl
-                (hash, Odoc.Id.to_string unit.Odoc_unit.parent_id)
-            with
-            | Some p -> Promise.await p
-            | None ->
-                let p, r = Promise.create () in
-                Hashtbl.add tbl (hash, Odoc.Id.to_string unit.parent_id) p;
-                let result = compile_one compile_mod hash in
-                Promise.resolve r result;
-                result)
-          units
+          (fun (other_unit_name, other_unit_hash) ->
+            match compile_other other_unit_hash with
+            | Ok r -> Some r
+            | Error _exn ->
+                Logs.debug (fun m ->
+                    m
+                      "Error during compilation of module %s (hash %s, \
+                       required by %s)"
+                      other_unit_name other_unit_hash
+                      (Fpath.filename unit.input_file));
+                None)
+          deps
       in
-      let r =
+      let includes =
         List.fold_left
-          (fun acc x ->
-            match (acc, x) with
-            | Error _, _ -> acc
-            | Ok acc, Ok x -> Ok (x @ acc)
-            | Ok _, Error x -> Error x)
-          (Ok []) r
+          (fun acc (_lib, path) -> Fpath.Set.add path acc)
+          Fpath.Set.empty
+          (Odoc_unit.Pkg_args.compiled_libs unit.pkg_args)
       in
-      r
+      Odoc.compile ~output_dir:unit.output_dir ~input_file:unit.input_file
+        ~includes ~suppress_warnings:(not unit.enable_warnings)
+        ~parent_id:unit.parent_id;
+      (match unit.input_copy with
+      | None -> ()
+      | Some p -> Util.cp (Fpath.to_string unit.input_file) (Fpath.to_string p));
+      Atomic.incr Stats.stats.compiled_units
+    in
+    let rec compile_mod : string -> ('a list, [> `Msg of string ]) Result.t =
+     fun hash ->
+      try
+        let units = Util.StringMap.find hash all_hashes in
+        let r =
+          Fiber.List.map
+            (fun unit ->
+              match
+                Hashtbl.find_opt tbl
+                  (hash, Odoc.Id.to_string unit.Odoc_unit.parent_id)
+              with
+              | Some p ->
+                  Promise.await p;
+                  None
+              | None ->
+                  let p, r = Promise.create () in
+                  Hashtbl.add tbl (hash, Odoc.Id.to_string unit.parent_id) p;
+                  let _result = compile_one compile_mod unit in
+                  Promise.resolve r ();
+                  Some unit)
+            units
+        in
+        Ok (List.filter_map Fun.id r)
+      with Not_found ->
+        Error (`Msg ("Module with hash " ^ hash ^ " not found"))
```

Comment:
I think it's acceptable to use exceptions here, but the exception handling seems quite for from where it can be raised.

Would it be possible to use `StringMap.find_opt`, turn the option into a result and use a `let*`?